File: day4/day4.rb
lines = File.readlines('day4.txt').map(&:chomp).map { |l| l.split(',') }

def line_to_range_arrays(l)
  first = l[0].split('-')
  second = l[1].split('-')
  [(first[0]..first[1]).to_a, (second[0]..second[1]).to_a]
end

# Part A
puts("Part A: #{lines.filter do |l|
  ranges = line_to_range_arrays(l)
  common = ranges[0] & ranges[1]
  common == ranges[0] || common == ranges[1]
end.count}")

# Part B
puts("Part B: #{lines.filter do |l|
  ranges = line_to_range_arrays(l)
  (ranges[0] & ranges[1]).any?
end.count}")

File: day4/day4.ex
lines =
  File.read!("day4.txt")
  |> String.split("\n", trim: true)
  |> Enum.map(fn x ->
    String.split(x, ",")
  end)
  |> Enum.map(fn x ->
    Enum.map(x, fn y ->
      [a, b] = String.split(y, "-", trim: true)

      Range.new(String.to_integer(a), String.to_integer(b))
    end)
  end)

# Part A
Enum.filter(lines, fn x ->
  first =
    Enum.at(x, 0)
    |> MapSet.new()

  second =
    Enum.at(x, 1)
    |> MapSet.new()

  MapSet.subset?(first, second) || MapSet.subset?(second, first)
end)
|> Enum.count()
|> IO.inspect()

# Part B
Enum.filter(lines, fn x ->
  !Range.disjoint?(Enum.at(x, 0), Enum.at(x, 1))
end)
|> Enum.count()
|> IO.inspect()
